cpp/src/arrow/gpu/cuda_memory.cc

```diff
@@ -480,5 +485,21 @@ Result<uint8_t*> GetHostAddress(uintptr_t device_ptr) {
   return static_cast<uint8_t*>(ptr);
 }
 
+Result<std::shared_ptr<MemoryManager>> DefaultMemoryMapper(ArrowDeviceType device_type,
+                                                           int64_t device_id) {
+  switch (device_type) {
+    case ARROW_DEVICE_CPU:
+      return default_cpu_memory_manager();
+    case ARROW_DEVICE_CUDA:
+    case ARROW_DEVICE_CUDA_HOST:
+    case ARROW_DEVICE_CUDA_MANAGED: {
+      ARROW_ASSIGN_OR_RAISE(auto device, arrow::cuda::CudaDevice::Make(device_id));
```

Comment:
This is a little weird but caused by this PR.  My first thought, at looking at this, is that we are returning a dangling object since `device` will be destroyed almost immediately.

I think `arrow::cuda::CudaDevice::Make` should probably be `arrow::cuda::CudaDevice::Get` and return `const CudaDevice&` or `const CudaDevice*`.

Reply:
I'll look at this in a subsequent PR rather than make that change here since it'll probably have other effects elsewhere